### PR TITLE
Import localized name and summary when uploading new version

### PR DESCRIPTION
**Pull request by hiikezoe**
* [X] The change has been successfully run locally.
* [X] Add tests to cover the changes added in this PR.

Fixes below items in #7825 

- [X] Support importing all localizations of the name and summary.
- [X] Validate summary and name length so they conform to AMO restrictions.
- [X] They can be updated with a version update; the fields will need to be validated and updated if they change.

### CODE REVIEW

**Comment by hiikezoe** [comment]
> src/olympia/files/tests/test_utils_.py:818: in test_atomic_lock_lifetime
    assert not lock_attained2
E   assert not True

The test passes locally,  I don't quite understand what happened there.

**Comment by diox** [comment]
Probably an intermittent failure, I've restarted the test

**Comment by hiikezoe** [comment]
I had to modify test data and test_install.py.

File: src/olympia/files/utils.py [hunk 1]
```diff
@@ -830,6 +830,23 @@ def parse_xpi(xpi, addon=None, minimal=False):
...
+                    'Localized name (%s) should have fewer than 50 '
+                    'characters.')
+                raise forms.ValidationError(msg % locale)
+    if isinstance(translations['summary'], dict):
+        for locale, localized_summary in translations['summary'].items():
+            if len(localized_summary) > 500:
+                msg = ugettext(
+                    'Localized summary (%s) should have fewer than 500 '
+                    'characters.')
+                raise forms.ValidationError(msg % locale)
```
Comment (EnTeQuAk) [inline src/olympia/files/utils.py]: I might be blind but where are these limits coming from?

Comment (hiikezoe) [reply src/olympia/files/utils.py]: From here https://github.com/mozilla/addons-server/blob/0c879c8bd6d0968ac890b9876a0209db81b7c9fc/src/olympia/addons/forms.py#L446
And now I noticed it's for theme.  The value is different for extension. https://github.com/mozilla/addons-server/blob/0c879c8bd6d0968ac890b9876a0209db81b7c9fc/src/olympia/addons/forms.py#L140 . I don't know why..

File: src/olympia/versions/models.py [hunk 2]
```diff
@@ -144,14 +144,20 @@ def save(self, *args, **kw):
...
             raise VersionCreateError(
                 'Addon is Mozilla Disabled; no new versions are allowed.')
 
         if parsed_data is None:
             parsed_data = utils.parse_addon(upload, addon)
+
+        # Import localized names and summaries
+        parsed_data = Addon.resolve_webext_translations(parsed_data, upload)
+        addon.name = parsed_data['name']
+        addon.summary = parsed_data['summary']
```
[the remark this answers is not in the document]
Comment (hiikezoe) [reply src/olympia/versions/models.py]: Yeah,  makes sense.  I will add a waffle flag for this.

**Comment by hiikezoe** [comment]
> I'm also curious as how this handles localization mismatches. Like, if the listing has a set of localizations for the metadata that doesn't match the set of localizations in the add-on itself.

It gets merged.   I.e. if the package doesn't have localized string for existing locales,  the existing localized strings are preserved.  I added a test case for this.

**Comment by hiikezoe** [comment]
Note that I added a waffle flag named 'import-metadata-from-package' in the updated PR.  I might be missing something since I don't know much about waffle flag.  I just knew it this morning. :)

**Comment by jvillalobos** [comment]
> It gets merged. I.e. if the package doesn't have localized string for existing locales, the existing localized strings are preserved. I added a test case for this.

Okay, that makes sense if this PR doesn't cover the changes to devhub. However, when we stop supporting editable summary and name, I think the behavior should be different. If the listing has a locale that isn't in the package, the name and summary for the default locale (from the package) will override the name and summary (on AMO) for the missing locale.

We need to do this because those fields won't be editable and we want them to be in sync with what's in the package.

**Comment by EnTeQuAk** [comment]
I am going to close this for now, for various reasons:

This needs to tackle quite a few more places *before* we can actually import `summary` and `description` properly into the add-on.

* We have to think about what happens when a version gets disabled or deleted
* Which means that we want to store some of that manifest-imported information on an add-ons `version` instead on the add-on itself. That'll make the whole disable/delete/rollback process much more reasonable
* All that work has to be behind a waffle-flag and shouldn't have any impact on initial deploy because it'll need developer communication and training
* There is work on Devhub that will have to be implemented at the same time (more or less) to allow for editing to be switched off when the waffle flags are swapped on/off

**Comment by hiikezoe** [comment]
Thanks for the explanations with deep insight.  I did take a brief chat with my manager about this and we are probably going to postpone [web-ext submit feature](https://github.com/mozilla/web-ext/issues/804) (which depends on this).

FWIW, I also started working on importing icons. It generates all ADDON_ICON_SIZES icons from imported icons by resizing the nearest size icon.  It may (or may not) be useful when we implement this feature properly.  Anyway just for reference https://github.com/hiikezoe/addons-server/commit/7ee49e5014d291ef7de5d4181dbdb0b674bee492#diff-024d4f5b3b9838580700aeeb2926b775R295

**Pull request closed without merging**